crates/cext/src/circuit.rs

```diff
@@ -378,6 +381,59 @@ pub struct OpCounts {
     len: usize,
 }
 
+/// @ingroup QkCircuit
+/// Append an arbitrary unitary matrix to the circuit.
+/// The user passes a row-major array of interleaved real-imaginary pairs #[
+/// representing the unitary matrix.
+///
+/// @param curcuit      A pointer to the circuit to add the unitary to.
+/// @param num_qubits   The width of unitary gate.
+/// @param qubits       The pointer to array of quibit indices.
+/// The array length must be equal to num_qubits.
+/// @param matrix       The pointer to the row-major `(re, im)` doucle array///
+/// # Example
+///
+///
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_unitary(
+    circuit: *mut CircuitData,
+    num_qubits: u32,
+    qubits: *const u32,
+    matrix: *const f64,
+) -> ExitCode {
+    // SAFETY: Caller quarantees pointer validation, alignment
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
+
+    // Dimension of the unitart: 2^n
+    let dim = 1 << num_qubits;
+
+    // Copy (re, im) pairs to Complex64
+    let raw = unsafe { std::slice::from_raw_parts(matrix, dim * dim * 2) };
+    let data: Vec<Complex64> = raw
+        .chunks_exact(2)
+        .map(|x| Complex64::new(x[0], x[1]))
+        .collect();
+
+    // Build ndarray::Array2
+    let mat = Array2::from_shape_vec((dim, dim), data).expect("Invalid shape for unitary matrix");
+
+    // Build qubit slice
+    let qargs: Vec<Qubit> = (0..num_qubits)
+        .map(|idx| Qubit(unsafe { *qubits.add(idx as usize) }))
+        .collect();
```

Comment:
You should actually be able to do this without a copy and new allocation. A `Qubit` is a `u32` in memory so just casting the pointer to u32s as a pointer to Qubits should be fine (with the approriate safety comment of course :) 

If you cast `qubits` to a `Qubit` pointer and just get a slice from it you can pass that to `push_packed_operation`  I think something like this should work:

```suggestion
    let qargs: &[Qubit] = unsafe {
        std::slice::from_raw_parts(qubits as *const, num_qubits as usize)
    };
```